src/lem/README.md

```diff
@@ -0,0 +1,131 @@
+## Lurk Evaluation Model (LEM)
+
+The main purpose of LEM is to represent Lurk's step function *as data*, which can
+be interpreted or used to synthesize the circuit.
+
+### Pointer cycle
+
+The variables of a LEM are represented by *meta pointers*, which become bound to
+concrete pointers during interpretation. So a `MetaPtr` is a named reference to
+a `Ptr`.
+
+To synthesize the circuit, the `Ptr`s that come from interpretation are hashed
+and become `ZPtr`s, whose data can be allocated in the circuit to generate
+*allocated pointers* (`AllocatedPtr`).
+
+```mermaid
+graph TD;
+    M[MetaPtr] --> P[Ptr];
+    P --> Z[ZPtr];
+    Z --> A[AllocatedPtr];
+```
+
+### Check
+
+We statically check the following properties:
+
+* Static single assignments
+* Non-duplicated input labels
+* One return per LEM path
+* Assign first, use later
+* Ensure always return
+
+### LEM macro syntax
+
+A summary of LEM syntax:
+
+* `let`
+* `return`
+* `match_tag`
+* `match_symbol`
+* `hash`, `unhash`
+* `hide`, `open`
+
+# Static Analysis
+
+The first traversal of LEM is the static analysis phase, where we allocate hash slots already considering deduplication. 
+
+## STEP 1
+
+    * calculate hash slots for all virtual paths.
+
+# Interpretation
+
+The second traversal occurs during interpretation of LEM. Given as input an expression, an environment and a continuation, the concrete path is followed by the interpreter, computing the output. 
+
+## STEP 2
+
+* calculate all preimages and images for all virtual paths.
+
+# Synthesis
+
+Before traversal we preallocate variables that were visited during interpretation, such as output variables and all preimages and images for the hash slots. Also before traversal we constrain all hash slots.
+
+As we traverse LEM, we push operations to a stack. As we pop from stack, we synthesize each operation, possibly creating new allocated pointers, which are inserted in the hash map of allocations. Those variables can be accessed later, avoiding unnecessary copies inside the circuit. We use an allocation manager component to avoid allocating constants multiples times.
```

Comment:
What benefit does this manual stack offer over the implicit stack a recursive traversal would provide?

Reply:
*Maybe* some speed, depending on how Rust compiles code. But that's just how I started coding this. Gabriel has plans to change this with a more structured definition for LEM

Reply:
The benefits of the manual stack is that it can grow as much as possible and that it is already TCO, since a "call" in tail position is a pop followed by a push (of course the contents of the manual stack are copied over to the "real" stack when popping, which allows us to set up a new manual frame using its values on top of the manual stack, then the real stack gets "popped" at the end of the loop block).

However, I would argue that using the real stack is better. Manual stack requires heap allocation, which requires syscalls to set up, which are slow because of the context switches. Furthermore, the heap is a complicated structure, so allocation is also more complex and more slow. *Furthermore*, heap allocations tend to lead to more cache misses. The only downside to the real stack, really, is that it is limited.